src/Configuration/Resolver/TaskConfigResolver.php

```diff
@@ -28,14 +36,21 @@ public function listAvailableTaskNames(): array
         return array_keys($this->taskMap);
     }
 
-    public function resolve(string $taskName, array $config): array
+    public function resolve(string $taskName): TaskConfig
     {
         $resolver = $this->fetchByName($taskName);
 
-        // Make sure metadata is never a part of the task configuration
+        $config = $this->taskMap[$taskName]['config'] ?? [];
+        $metadata = new Metadata($config['metadata'] ?? []);
+
         unset($config['metadata']);
+        $resolvedConfig = $resolver->resolve($config);
 
-        return $resolver->resolve($config);
+        return new TaskConfig(
+            $taskName,
+            $resolvedConfig,
+            $metadata
+        );
```

Comment:
`TaskConfg` instances should be kept in memory, non ?
Or instead of having such "TaskConfig resolver", you can:
- create a service definition for each TaskConfig in the container named like `test_config.<task_name>`.
- inject this service into the ones which need it

PS: The logic behind tasks configuration could be simpler. What's the reason of so many steps and layers ?

Reply:
See my answer here: https://github.com/phpro/grumphp/pull/1177#discussion_r2213475750